flang/lib/Lower/OpenMP.cpp

```diff
@@ -260,13 +254,101 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+      orderedClauseOperand, orderClauseOperand;
+  const auto &wsLoopOpClauseList = std::get<Fortran::parser::OmpClauseList>(
+      std::get<Fortran::parser::OmpBeginLoopDirective>(loopConstruct.t).t);
+  if (llvm::omp::OMPD_parallel_do ==
+      std::get<Fortran::parser::OmpLoopDirective>(
+          std::get<Fortran::parser::OmpBeginLoopDirective>(loopConstruct.t).t)
+          .v)
+    createParallelOp<Fortran::parser::OmpBeginLoopDirective>(
+        converter, eval,
+        std::get<Fortran::parser::OmpBeginLoopDirective>(loopConstruct.t));
+  else {
+    for (const auto &clause : wsLoopOpClauseList.v) {
+      if (const auto &privateClause =
+              std::get_if<Fortran::parser::OmpClause::Private>(&clause.u)) {
+        const Fortran::parser::OmpObjectList &ompObjectList = privateClause->v;
+        genObjectList(ompObjectList, converter, privateClauseOperands);
+      } else if (const auto &firstPrivateClause =
+                     std::get_if<Fortran::parser::OmpClause::Firstprivate>(
+                         &clause.u)) {
+        const Fortran::parser::OmpObjectList &ompObjectList =
+            firstPrivateClause->v;
+        genObjectList(ompObjectList, converter, firstPrivateClauseOperands);
+      }
+    }
+  }
+  for (const auto &clause : wsLoopOpClauseList.v) {
+    if (const auto &lastPrivateClause =
+            std::get_if<Fortran::parser::OmpClause::Lastprivate>(&clause.u)) {
+      const Fortran::parser::OmpObjectList &ompObjectList =
+          lastPrivateClause->v;
+      genObjectList(ompObjectList, converter, lastPrivateClauseOperands);
+    }
+  }
+  auto &doStmtEval = eval.getFirstNestedEvaluation();
+  auto *doStmt = doStmtEval.getIf<Fortran::parser::NonLabelDoStmt>();
+  const auto &loopControl =
+      std::get<std::optional<Fortran::parser::LoopControl>>(doStmt->t);
+  if (const auto *bounds =
+          std::get_if<Fortran::parser::LoopControl::Bounds>(&loopControl->u)) {
+    Fortran::lower::StatementContext stmtCtx;
+    lowerBound.push_back(fir::getBase(converter.genExprValue(
+        *Fortran::semantics::GetExpr(bounds->lower), stmtCtx)));
+    upperBound.push_back(fir::getBase(converter.genExprValue(
+        *Fortran::semantics::GetExpr(bounds->upper), stmtCtx)));
+    step.push_back(fir::getBase(converter.genExprValue(
+        *Fortran::semantics::GetExpr(bounds->step), stmtCtx)));
+  }
+  auto wsLoopOp = firOpBuilder.create<mlir::omp::WsLoopOp>(
+      currentLocation, lowerBound, upperBound, step, privateClauseOperands,
+      firstPrivateClauseOperands, lastPrivateClauseOperands, linearVars,
+      linearStepVars, scheduleClauseOperand.dyn_cast_or_null<StringAttr>(),
+      scheduleChunkClauseOperand,
+      collapseClauseOperand.dyn_cast_or_null<IntegerAttr>(),
+      noWaitClauseOperand.dyn_cast_or_null<UnitAttr>(),
+      orderedClauseOperand.dyn_cast_or_null<IntegerAttr>(),
+      orderClauseOperand.dyn_cast_or_null<StringAttr>());
```

Comment:
That's a long argument list. Has a single attribute dictionary argument been considered to slim this down a bit?

Reply:
That can done, however in `omp` lowering we follow this long list approach(for instance parallelOp also follows this). One benefit of this approach is that user/reader doesn't have to look anywhere else for arguments since `Op` is constructed there it self(using explicit arguments).